text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,667 @@
+several challenges. Messaging systems allow for asynchronous workflows, which
+means that the stages of producing and consuming a message can be separated by
+a considerable time gap (this can be minutes, hours, or days). Furthermore, one
+cannot rely on consistent instrumentation across all parts of the system that
+touch a message. Correlating producer and consumer stages are expected even when
+the intermediary forwarding the message between them is not instrumented.
+Finally, batching of messages can happen in many different parts of a message
+processing workflow, be it batch publishing, batch receiving, batch processing,
+or batch settling.
+
+Despite all those challenges, requirements for instrumentation of messaging
+scenarios are high. Besides correlating spans that model the different
+processing stages of a message, it should also be possible to determine the
+end-to-end latency of processing a message. If intermediaries are not
+instrumented, this shouldn't impact the correlation of producer and consumer
+stages. If, on the other hand, intermediaries are instrumented, spans from
+intermediary instrumentation should seamlessly integrate with producer and
+consumer instrumentation. This integration should not require any changes in
+producer or consumer instrumentation, and it should not cause any changes to
+the relationships of producer and consumer spans. Furthermore, it should be
+possible to provide tracing instrumentation as an out-of-the-box experience
+from messaging SDKs, without requiring any additional custom instrumentation
+from the user.
+
+This OTEP aims at proposing consistent guidelines for creating spans that model
+the stages of the messaging flow, and correlating those in a way so that the
+requirements sketched above can be met in a consistent way across messaging
+scenarios and different messaging systems.
+
+## Stages of producing and consuming messages
+
+As previously described in [OTEP 173](https://github.com/open-telemetry/oteps/blob/main/text/trace/0173-messaging-semantic-conventions.md#scenarios.),
+producing and consuming a message involves five stages:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  CR[Create] --> PU[Publish]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  end
+  subgraph CONSUMER
+  direction TB
+  RE[Receive] --> PR[Process]
+  PR --> SE[Settle]
+  end
+  PU --> INTERMEDIARY
+  INTERMEDIARY --> RE
+  SE -..- INTERMEDIARY
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget scenarios, or when
+   settlement happens on the broker), the settlement stage does not exist.
+
+The semantic conventions below define how to model those stages with spans.
+
+## Trace structure
+
```

Comment:
```suggestion
## Trace structure

This section outlines the general structure of messaging traces on a high level. The details are later
described in [Proposed changes and additions to the messaging semantic conventions](#proposed-changes-and-additions-to-the-messaging-semantic-conventions) (note especially the [Span relationships](#span-relationships) subsection).

```

Reply:
I hope this paragraph in the "Proposed changes" section makes things clearer:

> This section contains a concise and normative definition of what was outlined in the [Trace structure](#trace-structure) section. The following subsections are supposed to be merged into the semantic conventions, whereas the detailed description in the [Trace structure](#trace-structure) section in this OTEP will serve as explanation and future reference.